Added README.md for demo_cpp_nodes (#599)
ros2/demos: demo_nodes_cpp/README.md

2. `listener_serialized_message` 
3. `reuse_timer`
4. `add_two_ints_client_async` 
5. `list_parameters` 
6. `set_and_get_parameters`
7. `add_two_ints_server` 
8. `list_parameters_async` 
9. `set_and_get_parameters_async`
10. `allocator_tutorial` 
11. `one_off_timer` 
12. `set_parameters_callback`
13. `content_filtering_publisher` 
14. `parameter_blackboard` 
15. `talker`
16. `content_filtering_subscriber`
17. `parameter_event_handler`
18. `talker_loaned_message`
19. `even_parameters_node` 
20. `parameter_events`
21. `talker_serialized_message`
22. `listener`
23. `parameter_events_async` 
24. `listener_best_effort`

## **Build**

Run the command below to compile the `demo_nodes_cpp` ROS 2 package:

```bash
colcon build --packages-up-to demo_nodes_cpp
```

## **Run**

### Basic Talker & Listener

This runs a basic ROS 2 publisher and subscriber that exchanges the following string message with an incrementing integer:

> Hello World: <count_>

#### Talker

```bash
# Open new terminal
ros2 run demo_nodes_cpp talker
```

#### Listener [Default]

```bash
# Open new terminal
ros2 run demo_nodes_cpp listener
```

![](img/talker_listener.png)

#### Listener [Best Effort]

> Compared to **Listener [Default]**, `listener_best_effort` runs a ROS2 subscribe node that sets the Quality of Service (QoS) Reliability setting to **Best Effort** via the use of **rclcpp::SensorDataQoS**, as opposed to the default **Reliable**. Messages sent using this policy configuration attempts to deliver samples but may lose them if the network is not robust.

Reviewer: ```suggestion
Compared to **Listener [Default]**, `listener_best_effort` runs a ROS 2 subscriber node that sets the Quality of Service (QoS) Reliability setting to **Best Effort** via the use of **rclcpp::SensorDataQoS**, as opposed to the default **Reliable**.
Messages sent using this policy configuration attempts to deliver samples but may lose them if the network is not robust.
```

Author: Done.

```bash
# Open new terminal
ros2 run demo_nodes_cpp add_two_ints_client
```

#### Client [Asynchronous]

```bash
# Open new terminal
ros2 run demo_nodes_cpp add_two_ints_client_async
```

![](img/server_client.png)

### One-Off Timer

This runs `one_off_timer` that runs a periodic timer callback that **cancels** and **creates** a **Wall Timer** every **3 callbacks**.

```bash
ros2 run demo_nodes_cpp one_off_timer 
```

![](img/one_off_timer.png)

### Reuse Timer

Similar to previous, this runs `reuse_timer` that runs a periodic timer callback that **reuses the same Wall Timer** every **3 callbacks**.

```bash
ros2 run demo_nodes_cpp reuse_timer
```

### Serialized Messaging

This runs `talker_serialized_message` ROS 2 node that publishes a manual CDR serialization of the same string message.

```bash
# Open new terminal
ros2 run demo_nodes_cpp talker_serialized_message
```

This runs `listener_serialized_message` ROS2 2 node that subscribes and prints out the serialized string message published by `talker_serialized_message`. 
```bash
# Open new terminal
ros2 run demo_nodes_cpp listener_serialized_message
```

![](img/serialized_messaging.png)

### Content-Filter Messaging

> This runs `content_filtering_subscriber` and `content_filtering_publisher` ROS 2 nodes which exchanges the emergency temperature data and only filters out uninteresting data so that the subscription is not called.

Reviewer: ```suggestion
This runs `content_filtering_subscriber` and `content_filtering_publisher` ROS 2 nodes which exchanges temperature data on the `/temperature` topic.
However the subscriber requests that data is only sent if the temperature is less than -30 C or greater than 100 C, saving bandwidth.
```

Author: Done.

```bash
# Open new terminal
ros2 run demo_nodes_cpp content_filtering_subscriber
```

```bash
# Open new terminal
ros2 run demo_nodes_cpp content_filtering_publisher
```

![](img/content_filtering_messaging.png)

### List Parameters

> This runs `list_parameters` ROS 2 node which simply programmatically list example parameter names and prefixes:

Reviewer: ```suggestion
This runs `list_parameters` ROS 2 node which simply programmatically list example parameter names and prefixes.
```

Author: Done.

#### Synchronous

```bash
# Open new terminal
ros2 run demo_nodes_cpp list_parameters
```

#### Asynchronous

```bash
# Open new terminal
ros2 run demo_nodes_cpp list_parameters_async
```

### Set & Get Parameters

> This runs `set_and_get_parameters` ROS 2 node which programmatically sets and gets parameters:

Reviewer: ```suggestion
This runs `set_and_get_parameters` ROS 2 node which programmatically sets and gets parameters.
```

Author: Done.

#### Synchronous

```bash
# Open new terminal
ros2 run demo_nodes_cpp set_and_get_parameters
```

#### Asynchronous

```bash
# Open new terminal
ros2 run demo_nodes_cpp set_and_get_parameters_async
```

### Allocator Tutorial

> This runs `allocator_tutorial` ROS 2 node that publishes a `std_msgs/msg/UInt32` message that contains an integer representing the number of allocations and deallocations that happened during the program:

Reviewer: ```suggestion
This runs `allocator_tutorial` ROS 2 node that publishes a `std_msgs/msg/UInt32` message that contains an integer representing the number of allocations and deallocations that happened during the program.
```

Author: Done.

```bash
# Open new terminal
ros2 run demo_nodes_cpp allocator_tutorial
```
![](img/allocator_tutorial.png)

### Parameter Events

This runs `parameter_events`/`parameters_events_async` ROS2 node(s) which initiates 10 parameter events which changes an example string parameter.

> foo -> bar -> baz -> foobar -> foo -> bar -> baz -> foobar -> foo -> bar

#### Synchronous

```bash
# Open new terminal
ros2 run demo_nodes_cpp parameter_events
```

#### Asynchronous

```bash
# Open new terminal
ros2 run demo_nodes_cpp parameters_events_async
```

### Even Parameters Node

> This runs `even_parametes_node` ROS 2 node which shows a parameter callback that rejects all parameter update except tor those that set an even integer:

Reviewer: ```suggestion
This runs `even_parametes_node` ROS 2 node which shows a parameter callback that rejects all parameter update except tor those that set an even integer.
```

Author: Done.

```bash
# Open new terminal
ros2 run demo_nodes_cpp even_parameters_node
```

![](img/even_parameters_node.png)

### Set Parameters Callback

> This runs `even_parameters_callback` ROS 2 node which triggers a callback when ROS 2 double parameter `param1` is set. The callback then sets ROS 2 double parameter `param2` to a fixed `4.0`:

Reviewer: ```suggestion
This runs `set_parameters_callback` ROS 2 node which triggers a callback when ROS 2 double parameter `param1` is set.
The callback then sets ROS 2 double parameter `param2` to a fixed `4.0`.
```

Author: Done.

```bash
# Open new terminal
ros2 run demo_nodes_cpp set_parameters_callback
```

### Parameter Blackboard

> This runs `parameter_blackboard` ROS 2 node which instantiates a ROS2 parameter server which acts as "blackboard" which all nodes could get and set parametes value:

Reviewer: ```suggestion
This runs `parameter_blackboard` ROS 2 node which instantiates a ROS 2 parameter server which acts as a global "blackboard" for all nodes to get and set parameters.
```

Author: Done.

## **What Is This?**

> This demo provides different minimal examples of varying inter-node communcations in ROS 2.

Reviewer 2: ```suggestion
This package provides several examples that demonstrate various inter-node communcations in ROS 2.
```

Author: Done.

> This package consists of:

Reviewer 2: ```suggestion
This package consists of the following examples:
```

Author: Done.

19. `even_parameters_node` 
20. `parameter_events`
21. `talker_serialized_message`
22. `listener`
23. `parameter_events_async` 
24. `listener_best_effort`

## **Build**

Run the command below to compile the `demo_nodes_cpp` ROS 2 package:

```bash
colcon build --packages-up-to demo_nodes_cpp
```

## **Run**

### Basic Talker & Listener

This runs a basic ROS 2 publisher and subscriber that exchanges the following string message with an incrementing integer:

> Hello World: <count_>

#### Talker

```bash
# Open new terminal
ros2 run demo_nodes_cpp talker
```

#### Listener [Default]

```bash
# Open new terminal
ros2 run demo_nodes_cpp listener
```

![](img/talker_listener.png)

#### Listener [Best Effort]

Compared to **Listener [Default]**, `listener_best_effort` runs a ROS 2 subscriber node that sets the Quality of Service (QoS) Reliability setting to **Best Effort** via the use of **rclcpp::SensorDataQoS**, as opposed to the default **Reliable**.

Messages sent using this policy configuration attempts to deliver samples but may lose them if the network is not robust.

```bash
# Open new terminal
ros2 run demo_nodes_cpp listener_best_effort
```

### Basic Server & Client

> This runs a ROS 2 server that provides the service to processes two integers and outputs its sum back to ROS 2 client node. 

Reviewer 2: ```suggestion
This runs a ROS 2 server that provides a service to process two integers, outputting the sum back to ROS 2 client node. 
```

Author: Done.

#### Server

```bash
# Open new terminal
ros2 run demo_nodes_cpp add_two_ints_server
```

#### Client [Synchronous]

```bash
# Open new terminal
ros2 run demo_nodes_cpp add_two_ints_client
```

#### Client [Asynchronous]

```bash
# Open new terminal
ros2 run demo_nodes_cpp add_two_ints_client_async
```

![](img/server_client.png)

### One-Off Timer

This runs `one_off_timer` that runs a periodic timer callback that **cancels** and **creates** a **Wall Timer** every **3 callbacks**.

```bash
ros2 run demo_nodes_cpp one_off_timer 
```

![](img/one_off_timer.png)

### Reuse Timer

> Similar to previous, this runs `reuse_timer` that runs a periodic timer callback that **reuses the same Wall Timer** every **3 callbacks**.

Reviewer 2: ```suggestion
Similar to the previous demo, `reuse_timer` runs a periodic timer callback that **reuses the same Wall Timer** every **3 callbacks**.
```

Author: Done.

# Open new terminal
ros2 run demo_nodes_cpp list_parameters_async
```

### Set & Get Parameters

This runs `set_and_get_parameters` ROS 2 node which programmatically sets and gets parameters.

#### Synchronous

```bash
# Open new terminal
ros2 run demo_nodes_cpp set_and_get_parameters
```

#### Asynchronous

```bash
# Open new terminal
ros2 run demo_nodes_cpp set_and_get_parameters_async
```

### Allocator Tutorial

This runs `allocator_tutorial` ROS 2 node that publishes a `std_msgs/msg/UInt32` message that contains an integer representing the number of allocations and deallocations that happened during the program.

```bash
# Open new terminal
ros2 run demo_nodes_cpp allocator_tutorial
```
![](img/allocator_tutorial.png)

### Parameter Events

This runs `parameter_events`/`parameters_events_async` ROS2 node(s) which initiates 10 parameter events which changes an example string parameter.

> foo -> bar -> baz -> foobar -> foo -> bar -> baz -> foobar -> foo -> bar

#### Synchronous

```bash
# Open new terminal
ros2 run demo_nodes_cpp parameter_events
```

#### Asynchronous

```bash
# Open new terminal
ros2 run demo_nodes_cpp parameters_events_async
```

### Even Parameters Node

> This runs `even_parameters_node` ROS 2 node which shows a parameter callback that rejects all parameter update except tor those that set an even integer.

Reviewer 2: ```suggestion
This runs `even_parameters_node` ROS 2 node which shows a parameter callback that rejects all parameter updates except those that set an even integer.
```

Author: Done.
